### PR TITLE
[MLPERF] Retinanet

### PR DESCRIPTION
**Pull request by reddyn12**
TODO:
- [ ] Add bfloat16 support
- [ ] Speed up validation loop (multiproc)
- [x] Combine anchor gen code
- [ ] figure out unsynced or synced BN for speed
- [ ] fix synced BN train

REQUIREMENTS:
- numpy
- tqdm
- wandb
- boto3
- pandas
- pycocotools

TRAIN:
```./examples/mlperf/training_submission_v4.1/tinycorp/benchmarks/retinanet/implementations/tinybox_green/retina_train_green.sh```

### CODE REVIEW

**Comment by reddyn12** [comment]
Beam 3 BS 52 OOM
Beam 3 BS 32 1.3s but 37, 28, 28, 28 gb used in gpu
This is still proportional to BEAM 2 BS 52 step 2s

**Comment by chenyuxyz** [comment]
to lock the bounty, show a wandb link or equivalent that this trains to mlperf required target in 24 hours

**Comment by reddyn12** [comment]
School is on a training run rn, no available compute. Can you do a run on a tiny tn for atleast 1 epoch? 
Expected scores:
Epoch 1: 0.23
Epoch 2: 0.29
Epoch 3: 0.334989
Epoch 4: 0.3463

**Comment by reddyn12** [comment]
@chenyuxyz can you run to get a step-time for me to compare to?

**Comment by schlimeszn** [comment]
getting NANs, need to bisect

**Comment by schlimeszn** [comment]
Current WINO(1.2s step):
MODEL=retinanet CUDA=1 BS=32 GPUS=4 DEFAULT_FLOAT=float32 PYTHONPATH=. WANDB=1 TRAIN_BEAM=3 IGNORE_JIT_FIRST_BEAM=1 BEAM_COMPARE=0 BEAM_TIMEOUT_SEC=240 UOPS_REWRITE=0 BEAM_UOPS_MAX=0 BEAM_MIN_PROGRESS=20 WINO=1 LAZYCACHE=0 python3 examples/mlperf/model_train.py

**Comment by geohot** [comment]
Has this succeeded in producing a qualifying model?

**Comment by reddyn12** [comment]
After the merge, getting NAN. Doing a run on a pre merge branch: https://wandb.ai/schlimeszn/RetinaNet/runs/0xl12h6e?nw=nwuserschlime. It's on epoch 3/4 and the first 2 epochs match Nvidia's submission. This is fp32, batch 32, 1.2s step time, 12h epoch time. On the same branch I tried fp16, batch 64, 1.1s step time, 6 hr epoch time. The problem with this is I think I have gradient underflow(or resnet load not liking the casts) as the loss isn't going down properly: https://wandb.ai/schlimeszn/RetinaNet/runs/yo1kg8ou?nw=nwuserschlime

File: examples/mlperf/dataloader.py [hunk 1]
```diff
@@ -348,7 +348,209 @@ def shuffle_indices(file_indices, seed=None):
...
+
+  receive_batch = receive_batch_val if val else receive_batch_train
+  q_in, q_out = Queue(), Queue()
+
+  sz = (batch_size*BATCH_COUNT, 800, 800, 3)
+  if os.path.exists("/dev/shm/retinanet_X"): os.unlink("/dev/shm/retinanet_X")
+  shm = shared_memory.SharedMemory(name="resnet_X", create=True, size=prod(sz))
+  if not val:
+    bsz = (batch_size*BATCH_COUNT, 120087, 4)
+    if os.path.exists("/dev/shm/retinanet_YB"): os.unlink("/dev/shm/retinanet_YB")
```
Comment (roelofvandijk) [inline examples/mlperf/dataloader.py:496]: Consider using `pathlib`, `Path("/dev/shm/retinanet_YB").unlink(missing_ok=True)`

File: extra/datasets/openimages.py [hunk 2]
```diff
@@ -60,13 +61,42 @@ def openimages(subset: str):
...
+  return files
+
+def get_train_data():
+  with open(BASEDIR / 'train/train_data.json') as f:
+    data = json.load(f)
+  return data
+
+# @functools.lru_cache(None)
+def get_val_files():
+  if not (files:=glob.glob(p:=str(BASEDIR / "validation/data/*"))): raise FileNotFoundError(f"No validation files in {p}")
```
Comment (roelofvandijk) [inline extra/datasets/openimages.py:86]: Pathlib.glob https://docs.python.org/3/library/pathlib.html#pathlib.Path.glob

File: extra/datasets/openimages.py [hunk 3]
```diff
@@ -60,13 +61,42 @@ def openimages(subset: str):
...
+  return files
+
+def get_val_data():
+  with open(BASEDIR / 'validation/labels/openimages-mlperf.json') as f:
```
Comment (roelofvandijk) [inline extra/datasets/openimages.py:90]: `with (BASEDIR / 'validation/labels/openimages-mlperf.json').open() as f`

https://docs.python.org/3/library/pathlib.html#pathlib.Path.open

File: examples/mlperf/dataloader.py (hunk 1 above)
Comment (reddyn12) [reply examples/mlperf/dataloader.py]: thx, added both
